crates/server/src/authority/catalog.rs

```diff
@@ -654,7 +654,12 @@ async fn send_forwarded_response(
     };
 
     if can_validate_dnssec {
-        response_header.set_authentic_data(answers.dnssec_validated());
+        if answers.dnssec_validated() {
+            response_header.set_authentic_data(true);
+        } else if !request_header.checking_disabled() {
```

Comment:
Ah, there we go. So maybe this could/should go in the other commit?

This looks a little weird to me. Shouldn't the lookup already know whether it should have validated based on the input `CD` bit, and in that case, shouldn't `dnssec_validated()` always be false?

Reply:
well, it's complicated due to the way the RFC is worded.

the way I understood the RFC ([section 3.2.2 of RFC4035](https://datatracker.ietf.org/doc/html/rfc4035#section-3.2.2)) is that it does not make it a "MUST" that: if CD=1 then DNSSEC validation MUST *not* happen -- that's literally not in the RFC text.

what the RFC says (see quote below) is that the resolver SHOULD still return the answer (compiled from authoritative name servers) even if DNSSEC validation failed (e.g. outcome is Bogus or Indeterminate)

> When the CD bit is set, the recursive name server SHOULD, if possible, return the requested data to the originating resolver, even if the recursive name server's local authentication policy would reject the records in question.

then there's this part of that section (emphasis mine):

> If the CD bit is set, it indicates that the originating resolver is willing to perform whatever authentication its local policy requires.  **Thus, the resolver side of the recursive name server need not perform authentication on the RRsets in the response.**

I understand that as an *optional* optimization, not a requirement: if CD=1 then you MAY skip the DNSSEC validation and all the extra queries needed to perform that.

what other implementations do, and I have specifically inspected what `unbound` does at the wire level, is that they perform DNSSEC validation *regardless* of the state of the CD bit. the more surprising bit, is that `unbound` sets AD=1 *and* CD=1 in the response when the query has CD=1. (CD=1 in the response is expected; that's an RFC requirement). I'm not sure if the client can rely on that AD=1 in the CD=1 case because it specifically asked to disable checking of RRSIGs so :shrug:. I have not found a part of the RFC that explicitly states that AD=1 + CD=1 in the response is a valid state.

this implementation mimicks `unbound`'s interpretation of the RFC:

1. it performs DNSSEC validation regardless of the state of the CD bit
2. it allows AD=1+CD=1 in the response

I have however not added conformance tests for either of those behaviors (1 & 2). That is because we may want to adopt the MAY optimization I mentioned above and doing that would break tests that check for those behaviors (1 & 2)

Reply:
Sounds like you've carefully considered this at least! Maybe add some comments with some of this context in the code? The optional optimization seems like an obvious thing to do to me, I think it's valuable to avoid the extra work if a client indicates they don't need it.

Reply:
agreed. I think I agree with @japaric review of all of this. The RFCs are often not explicit in some cases, so I appreciate the fact that you reviewed unbound's behavior here. The docs that we added are good enough IMO.